File: Code/sorting_algorithms/quick_sort/quick_sort.cpp
// Quicksort algorithm by taking last element as pivot

#include <bits/stdc++.h>
using namespace std;

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partitionIndex(int *arr, int l, int r)
{
    int i = l - 1;
    int pivot = arr[r];

    for(int j = l; j < r; ++j)
    {
        if(arr[j] <= pivot)
        {
            ++i;
            swap(&arr[i], &arr[j]);
        }
        swap(&arr[i+1], &arr[r]);
    }
    return (i+1);
}

void quicksort(int *arr, int l, int r)
{
    int index;
    if(l < r)
    {
        index = partitionIndex(arr, l ,r);
        quicksort(arr, l, index-1);
        quicksort(arr, index+1, r);
    }
}

int main()
{
    //Sample input
    int arr[] = {106, 503, 444, 22, 90, 34, 108, 34, 73, 88, 66, 444, 10, 90};
    int n = sizeof(arr) / sizeof(int);

    quicksort(arr, 0, n-1);
    
    for(int i = 0; i < n; ++i)
        cout << arr[i] << " ";
    cout << endl;

    return 0;
}

File: Code/sorting_algorithms/quick_sort/quicksort.java
Quick sort in java

public class Main {

     static void quicksort(int arr [],int low,int high) {
         if (low >= high) {
             return;
         }
         int start = low;
         int end = high;
         int mid = start + (end - start) / 2;
         int pivot = arr[mid];
         while (start <= end) {
             while (arr[start] < pivot) {
                 start++;
             }
             while (arr[end] > pivot) {
                 end--;
             }
             if(start<=end){
                 int temp=arr[start];
                 arr[start]=arr[end];
                 arr[end]=temp;
                 start++;
                 end--;
             }
         }
         quicksort(arr,low,end);
         quicksort(arr,start,high);
     }
    public static void main(String[] args){
         int arr[]={5,6,2,1,7,4,8,9,12,3};
         quicksort(arr,0,arr.length-1);
     System.out.println(Arrays.toString(arr));
    }
}
